common/perfcounters.go

```diff
@@ -0,0 +1,130 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package common
+
+import (
+	"expvar"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/paulbellamy/ratecounter"
+)
+
+var perfMap map[string]*expvar.Map
+
+// PerfCounterIntRate describe a performance counter
+type PerfCounterIntRate struct {
+	count      expvar.Int
+	in         expvar.Int
+	rate       expvar.Int
+	rateMinute *ratecounter.RateCounter
+	avg        expvar.Float
+	avgMinute  *ratecounter.AvgRateCounter
+	startTime  time.Time
+}
+
+// from expvar go 1.7
+func expvarHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	fmt.Fprintf(w, "{\n")
+	first := true
+	expvar.Do(func(kv expvar.KeyValue) {
+		if !first {
+			fmt.Fprintf(w, ",\n")
+		}
+		first = false
+		fmt.Fprintf(w, "%q: %s", kv.Key, kv.Value)
+	})
+	fmt.Fprintf(w, "\n}\n")
+}
+
+// PerfcountersHandler return an http handler to show the values
+func PerfcountersHandler() http.Handler {
+	return http.HandlerFunc(expvarHandler)
+}
+
+// PerfHttpHanderFuncWrap is a simple wrapped to include call to Prolog/Epilog during each handler call
+func PerfHttpHanderFuncWrap(wrapped http.HandlerFunc, perf *PerfCounterIntRate) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		perf.Prolog(1)
+		wrapped(w, r)
+		perf.Epilog(1)
+	}
+}
+
+// NewPerfCounterIntPerMin create a new performance counter
+// count is the number of hit
+// in is the number of call within the Prolog/Epilog
+// permin is number of hit per minute
+// avg-ns is the time spent within the Prolog/Epilog per call
+func NewPerfCounterIntPerMin(name string) *PerfCounterIntRate {
+	i := strings.LastIndex(name, ".")
+	if i <= 0 {
+		panic("name must contain a dot, ex: api.counter")
+	}
+	mapName := name[:i]
+	cntName := name[i+1:]
+
+	p := &PerfCounterIntRate{}
+	p.rateMinute = ratecounter.NewRateCounter(60 * time.Second)
+	p.avgMinute = ratecounter.NewAvgRateCounter(60 * time.Second)
+
+	mroot := newOrGetPerfMap(mapName)
+	m := new(expvar.Map).Init()
+	mroot.Set(cntName, m)
+	m.Set("count", &p.count)
+	m.Set("in", &p.in)
+	m.Set("permin", &p.rate)
+	m.Set("avg-ns", &p.avg)
+	return p
+}
+
+// Prolog of the performance counter
+func (p *PerfCounterIntRate) Prolog(delta int64) {
+	p.count.Add(delta)
+	p.in.Add(delta)
+	p.rateMinute.Incr(delta)
+	p.rate.Set(p.rateMinute.Rate())
+	p.startTime = time.Now()
+}
+
+// Epilog of the performance counter
+func (p *PerfCounterIntRate) Epilog(delta int64) {
+	p.avgMinute.Incr(time.Since(p.startTime).Nanoseconds())
+	p.avg.Set(p.avgMinute.Rate())
+	p.in.Add(-delta)
+}
+
+func newOrGetPerfMap(name string) *expvar.Map {
+	if m, ok := perfMap[name]; ok {
+		return m
```

Comment:
shouldn't raise an error instead of returning the map ? if by mistake we use the same name ?

Reply:
good question, that's the default of expvar, but in our code (I don't have the example right now) we would like refer to the same map/counters (and it's legit) so that why I did that like this (newOrGet)